pkgs/http/lib/src/base_request.dart

```diff
@@ -89,14 +89,33 @@ abstract class BaseRequest {
   bool _finalized = false;
 
   static final _tokenRE = RegExp(r"^[\w!#%&'*+\-.^`|~]+$");
+
   static String _validateMethod(String method) {
     if (!_tokenRE.hasMatch(method)) {
       throw ArgumentError.value(method, 'method', 'Not a valid method');
     }
     return method;
   }
 
-  BaseRequest(String method, this.url)
+  /// On upload progress callback.
+  ///
+  /// If defined, this callback will be called when the upload progress changes.
+  ///
+  /// In browser, uses XMLHttpRequest's "xhr.upload.onLoad" event.
```

Comment:
More words may make this easier to read (to a limit, as always).

> /// When run in a browser, this callback relies on the
> /// browser's `XMLHttpRequest`'s `xhr.upload.onLoad` events.
> ///
> /// When run natively, the number of bytes is 
> /// .... 

I don't understand the description of the native version.
Probably because I don't know what stream we are talking about, and how you can possibly know the total length of a stream before it's done.

Reply:
`MultipartRequest` calculates the length value before the stream starts. If the length value is a non-calculated request (which I just added to the MultipartRequest), `onUploadProgress` will never be triggered.

Documentation will be extended

Reply:
@lrhn I actually prefer the original API (using a callback rather than a `Stream`) because it means that the `Client` implementation can easily check whether the caller is interesting in progress notification one at the beginning of their send implementation. In `package:cupertino_http`, I'd probably implement something like:

```dart
  @override
  Future<StreamedResponse> send(BaseRequest request) async {
    ...
    if (request.onUploadProgress != null) {
      // This `Timer` is an unnecessary expense if no one is listening for notifications.
      final timer = Timer.periodic(const Duration(milliseconds: 100), (timer) { 
      if (taskTracker.isDone) {
        timer.cancel();
      } else {
        progress = Progress(task.countOfBytesExpectedToSend,  task.countOfBytesSent);
        request.onUploadProcess(progress);
      });
    }
```

If the `Stream/StreamController` approach were used, it seems like the `StreamController` would have to be public and so the `send` implementation can do something like:

```
  @override
  Future<StreamedResponse> send(BaseRequest request) async {
    void setupTimer() {
    
    }

    if (request.controller.isListening) {
      setupTimer();
    } else {
      request.onListen = setupTimer;
   }
```